xen/arch/arm/coproc/coproc.c

```diff
@@ -73,162 +69,293 @@ static struct coproc_device *coproc_find_by_path(const char *path)
 }
 
 static struct vcoproc_instance *coproc_init_vcoproc(struct domain *d,
-                                                    struct coproc_device *coproc)
+                                                    struct mcoproc_device *mcoproc,
+                                                    const struct dt_device_node *np_vcoproc)
 {
     struct vcoproc_instance *vcoproc;
-    int ret;
+    int ret = 0;
+    int i;
 
     vcoproc = xzalloc(struct vcoproc_instance);
     if ( !vcoproc )
     {
         printk("Failed to allocate vcoproc_instance for %s\n",
-               dev_path(coproc->dev));
+               dev_path(mcoproc->dev));
         return ERR_PTR(-ENOMEM);
     }
 
-    vcoproc->coproc = coproc;
+    vcoproc->mcoproc = mcoproc;
     vcoproc->domain = d;
     vcoproc->state = VCOPROC_UNKNOWN;
     spin_lock_init(&vcoproc->lock);
 
-    ret = coproc->ops->vcoproc_init(vcoproc);
-    if ( ret )
+    vcoproc->num_mmios = dt_number_of_address(np_vcoproc);
+
+    if ( vcoproc->num_mmios != mcoproc->num_mmios )
     {
-        printk("Failed to initialize vcoproc_instance for %s\n",
-               dev_path(coproc->dev));
+        printk("MMIO num mistmatch for \"%s\", %u vs %u\n",
+               dt_node_full_name(np_vcoproc), vcoproc->num_mmios, mcoproc->num_mmios);
+        ret = -ENODEV;
         goto out;
     }
 
-    spin_lock(&coproc->vcoprocs_lock);
-    list_add(&vcoproc->vcoproc_elem, &coproc->vcoprocs);
-    spin_unlock(&coproc->vcoprocs_lock);
-
-    return vcoproc;
-
-out:
-    xfree(vcoproc);
-    return ERR_PTR(ret);
-}
+    vcoproc->mmios = xzalloc_array(struct vcoproc_mmio, vcoproc->num_mmios );
+    if ( !vcoproc->mmios )
+    {
+        printk("Failed to allocate %d mmio(s) for \"%s\"\n", vcoproc->num_mmios,
+               dev_path(mcoproc->dev));
+        ret = -ENOMEM;
+        goto out;
+    }
 
-struct vcoproc_instance *coproc_get_vcoproc(struct domain *d,
-                                            struct coproc_device *coproc)
-{
-    struct vcoproc_instance *vcoproc = NULL;
-    bool_t found = false;
-    unsigned long flags;
+    for ( i =0; vcoproc->num_mmios > i; i++)
```

Comment:
unfortunately xen coding style differs enough to let checkpatch.pl blaime each line